### ISSUE TITLE
Is there a hint for unqualified imports

### ISSUE THREAD

**Issue by phadej** [ndmitchell/hlint#769]
I'd like to prohibit unqualified imports without explicit import list:

```haskell
import Data.List                 -- bad
import qualified Data.List as L  -- ok
import Data.List (foldl')        -- ok
```

This should be configurable, e.g. to allow some modules to be imported unqualified:

```haskell
import Prelude hiding (null)
-- or
import Prelude.Compat -- should be configurable to be ok
```

**Comment by ndmitchell** [comment]
See https://github.com/ndmitchell/hlint#restricting-items for whats there now. I don't quite know the boundaries of its power, but I'm fine with patches making it do the things you describe.

**Comment by phadej** [comment]
@ndmitchell thanks for prompt reply. I actually use all functionality from that section, and it doesn't mention *unqualified imports*. I'll see if I can hack that together.

**Comment by phadej** [comment]
Bummer. Implicitly listing `Prelude` is fine, but implicitly listing all package own modules is cumbersome. I need to think, how to not make this a configuration misery.

**Comment by ndmitchell** [comment]
There is a ticket to make those patterns take wildcards. If that happened then you might expect a users package to mostly share a common prefix, so you could whitelist `Development.Shake.**` and that would do it.

**Comment by phadej** [comment]
I'd like to share the same `.hlint,yaml` for all my projects. Essentially, I'd like HLint to be able to get information from `.cabal` files about what modules are local and which aren't. But that seems to be too big a change.

**Comment by danbroooks** [comment]
I'd really like this feature, even if you have to define a namespace in hlint, so something like:

`MyApp.QualifiedModules.*` would be great

**Comment by ndmitchell** [comment]
Feel free to go for it. I imagine you need to add something to https://github.com/ndmitchell/hlint/blob/master/src/Hint/Restrict.hs and https://github.com/ndmitchell/hlint/blob/cf972fedcc6e31e94404cbcda6dfbb052b6b7659/src/Config/Type.hs#L28. You parse the config somehow, turn it into a RestrictedItem structure, then apply that. I don't have any particularly good idea how this might work, as personally I wouldn't be interested in using it. I'd first prototype possible .yaml file syntaxes, and their semantics, and then go from there.

**Comment by danbroooks** [comment]
I had a bit of a play around and I have something working, but like you say the key to this now is how to get it so the configuration is nice...

What I have so far uses a new field on `modules` called `qualifiedStyle`, so for example the config:

```
- modules: { name: [Data.List], as: L, qualifiedStyle: ['pre'] }
```

Would make it so that all imports of `Data.List` have to be:

```hs
import qualified Data.List as L
```

Is what we actually want here though, actually a hint that is on by default for all modules, and obviously can be `ignored` like any other hint, with the ability to switch it off on a per-module basis? So using @phadej 's example, you could have a config like so:

```
- ignore: { name: "Implicit unqualified imports"
          , within: 
            [ Prelude
            , "MyApp.UnqualifiedModules.*"
            ] 
          }
```

Or something? My initial thinking for my use case would be to force qualified imports of certain modules, but on re-reading the issue, I think going at it this way round would be the right way to go.

If there is no view either way, would you be happy with the above approach @ndmitchell , if I was to proceed with that and submit it as a PR?

**Comment by danbroooks** [comment]
Having said that, I'm not sure I myself would actually like the above configuration on in my project now that I have given it some though. 

My original use case that brought me here was to force a submodule to always use a qualified prefix, I would like it if my: `App.Queries.*` modules always had to be qualified, and with a specific qualified initial, so something like `DB`, that way it is easy to visually parse queries, as they will always be prefixed `DB.selectSomeStuff` etc.

I think there is a use case for both of these things to be honest, and it would be up to the user how they would want to configure things, however that does put a spanner in the works with how this would fit in with hlint and its configuration.

**Comment by ndmitchell** [comment]
I don't see adding wildcard logic to modules as a problem. I did originally think of having modules contain as/qualifiedStyle to gain consistency, so that seems a plausible design to me. I would potentially write `name: ["Data.ByteString", as: "BS", qualifiedStyle: "always"]}` in larger projects I was working on, so I can see the attraction. I wouldn't have qualifiedStyle pre vs post though - I think once a project switches to post qualified modules they will do so universally, and I think there's a GHC warning to help with that.

**Comment by shayne-fletcher** [comment]
> [...]
> I think once a project switches to post qualified modules they will do so universally, and I think there's a GHC warning to help with that.

That's right! It's `-Wprepositive-qualified-syntax`. Full details [here]( https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/import_qualified_post.html#extension-ImportQualifiedPost).
